src/cudamatrix/cu-kernels.cu

```diff
@@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
   }
 }
 
+template<typename Real>
+__global__
+static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
+                                  Real* obfj_terms, MatrixDim objf_dim, 
+                                  Real* obfj_derivs, MatrixDim derivs_dim) {
+  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
+  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
+  int32_cuda scores_index = i + j * scores_dim.stride;
+  Real K = 1.0 / (scores_dim.rows - 2.0);
+  Real L = scores[scores_index];
+  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
```

Comment:
I think the only not-100%-trivial thing about the derivatives is the fact
that different parts of the matrix have different scaling factors.  You
could probably compute the objf and derivs as follows using individual
kernels.

  - get matrix of scores.
  - apply fixed-scaling-1 to matrix of scores (to negate different-class)
  - compute soft-hinge function
  - Compute TraceMatMat of this matrix with a fixed scaling matrix
fixed-scaling-2 (with 1/(num-rows-2) for different-class members) to get
the objf
  - use the Sigmoid function to compute the derivative of the soft-hinge,
and dot-multiply
 (passing in the fixed-scaling-2 from the step above) to get the derivative
of the soft-hinge.
  - Multiply the derivatives by fixed-scaling-1 * fixed-scaling-2.  These
are the derivatives w.r.t. the raw scores.

There may be a few signs wrong here.
However, it would be more efficient to do all of the above in a single
kernel.
You can easily do it in the same kernel as computes the objective-function
terms.  [do summation via matrix-sum though].

Dan



On Sat, Feb 13, 2016 at 3:58 PM, david-ryan-snyder <notifications@github.com
> wrote:

> In src/cudamatrix/cu-kernels.cu
> <https://github.com/danpovey/kaldi/pull/5#discussion_r52833363>:
>
> > @@ -2094,6 +2095,26 @@ static void _diff_xent(const int32_cuda* vec_tgt, Real* mat_net_out, Real* vec_l
> >    }
> >  }
> >
> > +template<typename Real>
> > +__global__
> > +static void _compute_xvector_objf(const Real* scores, MatrixDim scores_dim,
> > +                                  Real* obfj_terms, MatrixDim objf_dim,
> > +                                  Real* obfj_derivs, MatrixDim derivs_dim) {
> > +  int32_cuda i = blockIdx.x * blockDim.x + threadIdx.x;
> > +  int32_cuda j = blockIdx.y * blockDim.y + threadIdx.y;
> > +  int32_cuda scores_index = i + j * scores_dim.stride;
> > +  Real K = 1.0 / (scores_dim.rows - 2.0);
> > +  Real L = scores[scores_index];
> > +  if (i < scores_dim.cols && j < scores_dim.rows && i < j) {
>
> After looking at it more, I don't think it's better to just do this in a
> cuda kernel.
>
> Also, I still need to make kernels for the actual derivatives, which are
> somewhat nontrivial to compute in an efficient way... I don't think it's
> possible to use Pegah's idea to handle them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/5/files#r52833363>.
>